src-electron/generator/helper-attribute.js

```diff
@@ -76,6 +79,89 @@ async function attributeDefault(options) {
   return templateUtil.templatePromise(this.global, p)
 }
 
+/**
+ * Given an attribute Id determine its corresponding atomic identifier from the
+ * atomic table.
+ * @param {*} attributeId
+ */
+async function as_underlying_atomic_identifier_for_attribute_id(attributeId) {
+  let attributeDetails =
+    await queryZcl.selectAttributeByAttributeIdAndClusterRef(
+      this.global.db,
+      attributeId,
+      null
+    )
+  let atomicInfo = attributeDetails
+    ? await queryZcl.selectAtomicType(
+        this.global.db,
+        [attributeDetails.packageRef],
+        attributeDetails.type
+      )
+    : null
+  // If attribute type directly points to the atomic type then return that
+  if (atomicInfo) {
+    // All types in the number and string table should be found here.
+    return atomicInfo.atomicId
+  } else {
+    // If attribute type does not point to atomic type
+    let dataType = await queryZcl.selectDataTypeByNameAndClusterId(
+      this.global.db,
+      attributeDetails.type,
```

Comment:
This whole block of code is a bit weird.... Mining in all these different places for a name: I can understand where it comes from, but shouldn't there be a better way to do this? Just one table "TYPES" that contains all known types, be it enum values or bitmap values or whatever?

I wonder if having separate tables for these was a mistake...

Reply:
All types are in the data_type table but that only has the common characteristics with the distinct attributes of each type going into their corresponding tables.